File: backend/app/models/tribe.py
from datetime import datetime
from sqlalchemy import Column, String, Integer, DateTime, ForeignKey, JSON, Boolean
from sqlalchemy.orm import relationship
from geoalchemy2 import Geography

from .base import Base

class Tribe(Base):
    __tablename__ = "tribes"

    id = Column(String, primary_key=True)
    name = Column(String, nullable=False)
    description = Column(String)
    image_url = Column(String)
    created_by = Column(String, ForeignKey("users.id"))
    status = Column(String, default="pending")  # pending, active, archived
    location = Column(Geography(geometry_type='POINT', srid=4326))
    cultural_focus = Column(JSON)  # List of cultural aspects
    traditions = Column(JSON)  # List of traditions
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    members = relationship("User", back_populates="current_tribe")
    events = relationship("TribeEvent", back_populates="tribe")
    votes = relationship("TribeVote", back_populates="tribe")

class TribeEvent(Base):
    __tablename__ = "tribe_events"

    id = Column(String, primary_key=True)
    tribe_id = Column(String, ForeignKey("tribes.id"))
    title = Column(String, nullable=False)
    description = Column(String)
    date = Column(DateTime, nullable=False)
    location = Column(String, nullable=False)
    image_url = Column(String)
    created_by = Column(String, ForeignKey("users.id"))
    attendees_count = Column(Integer, default=0)
    cultural_type = Column(String)
    traditions = Column(JSON)
    created_at = Column(DateTime, default=datetime.utcnow)

    # Relationships
    tribe = relationship("Tribe", back_populates="events")
    attendees = relationship("EventAttendee", back_populates="event")

class TribeVote(Base):
    __tablename__ = "tribe_votes"

    id = Column(String, primary_key=True)
    tribe_id = Column(String, ForeignKey("tribes.id"))
    title = Column(String, nullable=False)
    description = Column(String)
    created_by = Column(String, ForeignKey("users.id"))
    ends_at = Column(DateTime, nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow)

    # Relationships
    tribe = relationship("Tribe", back_populates="votes")
    options = relationship("VoteOption", back_populates="vote")

class VoteOption(Base):
    __tablename__ = "vote_options"

    id = Column(String, primary_key=True)
    vote_id = Column(String, ForeignKey("tribe_votes.id"))
    text = Column(String, nullable=False)
    votes_count = Column(Integer, default=0)

    # Relationships
    vote = relationship("TribeVote", back_populates="options")

File: pendo_mobile/lib/features/tribes/domain/entities/tribe.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:pendo_mobile/features/shared/models/location_model.dart';

part 'tribe.freezed.dart';
part 'tribe.g.dart';

@freezed
class TribeModel with _$TribeModel {
  const factory TribeModel({
    required String id,
    required String name,
    required String description,
    @JsonKey(name: 'image_url') required String imageUrl,
    @JsonKey(name: 'created_by') required String createdBy,
    required String status,
    @JsonKey(name: 'member_count') required int memberCount,
    @JsonKey(name: 'events_count') required int eventsCount,
    @JsonKey(name: 'messages_count') required int messagesCount,
    @JsonKey(name: 'chief_id') String? chiefId,
    @JsonKey(name: 'location') LocationModel? location,
    @JsonKey(name: 'cultural_focus') List<String>? culturalFocus,
    @JsonKey(name: 'traditions') List<String>? traditions,
    @JsonKey(name: 'events') List<TribeEventModel>? events,
    @JsonKey(name: 'created_at') required DateTime createdAt,
    @JsonKey(name: 'updated_at') required DateTime updatedAt,
  }) = _TribeModel;

  factory TribeModel.fromJson(Map<String, dynamic> json) =>
      _$TribeModelFromJson(json);
}

@freezed
class TribeEventModel with _$TribeEventModel {
  const factory TribeEventModel({
    required String id,
    required String title,
    required String description,
    required DateTime date,
    required String location,
    @JsonKey(name: 'image_url') String? imageUrl,
    @JsonKey(name: 'created_by') required String createdBy,
    @JsonKey(name: 'attendees_count') required int attendeesCount,
    @JsonKey(name: 'cultural_type') String? culturalType,
    @JsonKey(name: 'traditions') List<String>? traditions,
    @JsonKey(name: 'created_at') required DateTime createdAt,
  }) = _TribeEventModel;

  factory TribeEventModel.fromJson(Map<String, dynamic> json) =>
      _$TribeEventModelFromJson(json);
}

@freezed
class TribeVoteModel with _$TribeVoteModel {
  const factory TribeVoteModel({
    required String id,
    @JsonKey(name: 'tribe_id') required String tribeId,
    required String title,
    required String description,
    @JsonKey(name: 'created_by') required String createdBy,
    @JsonKey(name: 'options') required List<VoteOptionModel> options,
    @JsonKey(name: 'ends_at') required DateTime endsAt,
    @JsonKey(name: 'created_at') required DateTime createdAt,
  }) = _TribeVoteModel;

  factory TribeVoteModel.fromJson(Map<String, dynamic> json) =>
      _$TribeVoteModelFromJson(json);
}

@freezed
class VoteOptionModel with _$VoteOptionModel {
  const factory VoteOptionModel({
    required String id,
    required String text,
    @JsonKey(name: 'votes_count') required int votesCount,
  }) = _VoteOptionModel;

  factory VoteOptionModel.fromJson(Map<String, dynamic> json) =>
      _$VoteOptionModelFromJson(json);
}
